packages/server/src/adapters/ws.ts

```diff
@@ -300,6 +301,26 @@ export function applyWSSHandler<TRouter extends AnyRouter>(
       }
     });
 
+    const keepAliveMs = opts.keepAliveMs || 20_000;
+    let pingInterval: NodeJS.Timer | null = null;
+    let pongWait: NodeJS.Timeout | null = null;
+    if (keepAliveMs > 0) {
+      pingInterval = setInterval(() => {
+        if (client.readyState === 1 /* ws.OPEN */) {
+          pongWait = setTimeout(() => client.terminate(), keepAliveMs);
+
+          client.once('pong', () => {
+            if (pongWait) {
+              clearTimeout(pongWait);
+              pongWait = null;
+            }
+          });
+
+          client.ping();
+        }
+      }, keepAliveMs);
```

Comment:
I don't like `setInterval` with backpressure, wouldn't this e better implemented with a recursive `setInterval`? Then we could also measure time for each pog and use that for something later.


Mock code:

```ts

const pingPongInterval = 5_000;
const pingPongMaxLatency = pingPongInterval;

let pingTimeout = pingPongInterval ? setTimeout(ping, pingPongInterval) : null;
function ping() {
  const timeStart = Date.now();

  client.once('pong', onPong);
  const timeout = setTimeout(onTimeout, pingPongMaxLatency);

  function onPong() {
    // Clear up timeout so it doesn't terminate the connection
    clearTimeout(timeout);
    const latency = Date.now() - timeStart;
    const nextPing = pingPongInterval - latency;
    pingTimeout = setTimeout(ping, nextPing);
  }
  function onTimeout() {
    // Stop listening for pong to avoid memory leak
    client.off('pong', onPong);
    client.terminate();
  }
}

// in `client.once('close', () => { ... })`:
client.once('close', () => {
  // Clear up ping interval to avoid memory leaks
  pingTimeout && clearTimeout(pingTimeout);
  // [...]
});

```

Reply:
Alternatively, we can do as they suggest in the docs? https://github.com/websockets/ws#how-to-detect-and-close-broken-connections

This would be a bit less resource intensive, I think?

Reply:
Oh thanks for the hint, I am going to take a look at that and try to come up with a better implementation.